### PR TITLE
Tweak Git blame UI/UX

### LINKED ISSUE #38377: Move “git extras” (i.e. git blame etc.) extension to core workflow
- [x] https://github.com/sourcegraph/sourcegraph/issues/39433
- [x] https://github.com/sourcegraph/sourcegraph/issues/39437
- [x] https://github.com/sourcegraph/sourcegraph/issues/41314

### LINKED ISSUE #41314: Git blame UI review improvements
UI review for issue #40615. Necessary to close #38377 :

**Edge cases that @taras-yemets raised:**

- [x]  "Git blame column with Codecov line decorations line background on hover conflict ([Loom](https://www.loom.com/share/44bb0d3ec655455a9a730ffd4a61fd71)). This kind of conflict may happen with any extension which changes line colors. Customers with enableLegacyExtensions experimental feature flag enabled are affected."
  - As this only impacts customers with the legacy extensions enabled _and_ have to enable both CodeCov and blame, I think a minimal effort solution is best here. I think your first suggestion that blame highlight takes precedence over other extensions is sufficient, as it's our top-used extension based on our current data. 
- [x] "There is a gap between the blame decoration and the popover. When there are multiple lines in a row with decorations and the user hovers over one of them and then tries to hover over the popover, they should move the cursor over the code area, but not over the blame column, as the previous line decoration will trigger opening its popover. Moving the cursor over the code area may trigger showing the code-intel popover, so we may end up with two popovers ([Loom](https://www.loom.com/share/82364b7d2bf64043bfa07316ec207495))."
  - This is less an edge case, but poor UX of how the blame popover interacts with the popover. I'm [seeing that there's a latency](https://www.loom.com/share/330439ac77fd4750850e6ceed8f5f705) in the blame popover disappearing when a user's cursor has moved outside of the blame column, which is why there's an overlap with the code intel popover. Is it possible for us to instantly close the blame popover if a user's cursor moves outside the blame column? 

**Additional UI feedback** 

- [x] The blame popover still looks large when used alongside the code intel popover. Having trouble inspecting the font sizes of the popover using the inspect tool ☹️ , could someone double-check the styling?
Screen Shot 2022-09-05 at 8 18 48 PM
Screen Shot 2022-09-05 at 8 18 52 PM

- [x] You can also see in the above screenshots that the positioning of the popover is different from the code intel popover. The code intel popover is anchored directly above the hovered token, whereas our blame popover has an ambiguous relationship to the line being hovered. It's standard for popovers on Sourcegraph to be positioned directly alongside the content; could we adjust the positioning? 
- [x]  I've intentionally changed the font of the blame column from the font used for code to our body font family. Blame is content, not code, and it's more consistent with Github as well! 
- [x] On a similar note on font, the row being hovered should transition to [being Label/small-underline](https://www.figma.com/file/7bVINrmqfDpxmLx3gv1uVa/Sunsetting-extensions?node-id=2116%3A22361). 
- [x] (**Open to feedback on this**) I transitioned to using absolute dates for better formatting, but open to feedback on this!

/cc @muratsu @jjinnii @ryankscott

### PR DESCRIPTION
**Pull request by philipp-spiess**
Closes #41314
Closes #38377

This PR tweaks the UI/UX of the newly added migrated Git blame extension based on @jjinnii's feedback. I posted some comments on the [original issue](https://github.com/sourcegraph/sourcegraph/issues/41314) but will re-iterate them here as well:

> ⚠️  "Git blame column with Codecov line decorations line background on hover conflict ([Loom](https://www.loom.com/share/44bb0d3ec655455a9a730ffd4a61fd71)). This kind of conflict may happen with any extension which changes line colors. Customers with enableLegacyExtensions experimental feature flag enabled are affected."
As this only impacts customers with the legacy extensions enabled and have to enable both CodeCov and blame, I think a minimal effort solution is best here. I think your first suggestion that blame highlight takes precedence over other extensions is sufficient, as it's our top-used extension based on our current data.

Just to clarify with that you mean the behavior that is already shown in Tara's video?

I suppose this is already how it works (because of the video) but I have a hard time checking this since I don't really understand how to enable the codecov extension. Do you have it running by any chance @jjinnii and can test this on dot-com?

**TL;DR: I think we don't need to make any changes but this needs verification.**

> ✅ "There is a gap between the blame decoration and the popover. When there are multiple lines in a row with decorations and the user hovers over one of them and then tries to hover over the popover, they should move the cursor over the code area, but not over the blame column, as the previous line decoration will trigger opening its popover. Moving the cursor over the code area may trigger showing the code-intel popover, so we may end up with two popovers ([Loom](https://www.loom.com/share/82364b7d2bf64043bfa07316ec207495))."
This is less an edge case, but poor UX of how the blame popover interacts with the popover. I'm [seeing that there's a latency](https://www.loom.com/share/330439ac77fd4750850e6ceed8f5f705) in the blame popover disappearing when a user's cursor has moved outside of the blame column, which is why there's an overlap with the code intel popover. Is it possible for us to instantly close the blame popover if a user's cursor moves outside the blame column?

I don't think this would be a good fix UX wise. If you look at Tara's video again, all he's doing is to try and click the button on the popover which causes his mouse to move a bit to the center of the popover. I don't think it would be intended in this case to close the git blame popover in favor of code intel.

I've played around with this feature now as well and what's really unexpected for me is that when you are not super fast, moving the mouse to the content of the bit blame window will hover over another line and thus change the git blame info to the one on the line above/below.

I think what is really unexpected is that somehow if you hover over a new blame column, the popup will switch instantly but if you hover from a blame column to the code to show a code intel column, this is not the case.

However, I think this is less of an issue after one of the other changes below that changes the position of the git blame popup. Since the popup is now no longer shifted to the right, you don't need to move the mouse outside the blame column but simply upwards/downwards which prevents any interferences with the code intel popup.

**TL;DR: I recommend we keep this as-is since a behavior change here is not trivial and the changed horizontal offset of the popover patches over the most noticeable issue when moving the mouse outside the blame column**

> ✅ The blame popover still looks large when used alongside the code intel popover. Having trouble inspecting the font sizes of the popover using the inspect tool ☹️ , could someone double-check the styling?

I fixed this as good as possible. I needed to make changes to the paddings and the content text size. See the screnshot below.

@jjinnii Here's a trick that I use to make the inspect tool work with popovers. Paste the following snipped into the console:

```
setTimeout(() => { debugger; }, 2000);
```

After pressing enter, you have two seconds to move the mouse to the popover and then the debugger breakpoint will kick in and freeze the JavaScript and DOM mutations. You can then use the inspector tool however often you want 🙂 

**TL;DR: Fixed**

>  ✅ You can also see in the above screenshots that the positioning of the popover is different from the code intel popover. The code intel popover is anchored directly above the hovered token, whereas our blame popover has an ambiguous relationship to the line being hovered. It's standard for popovers on Sourcegraph to be positioned directly alongside the content; could we adjust the positioning?

I've changed this but I think I know the reason why it was added in the first place. Without it, you can't see any of the blame lines behind the popover. It's not too big a deal I think since you can always just move the mouse away again to get more context.

What I really like about this change though is that it causes you to no longer need to move the mouse into the content area if you want to click the link and thus it kind of works the second issue Taras raised.

**TL;DR: Fixed**

> ✅ On a similar note on font, the row being hovered should transition to [being Label/small-underline](https://www.figma.com/file/7bVINrmqfDpxmLx3gv1uVa/Sunsetting-extensions?node-id=2116%3A22361).

**TL;DR: Not an issue**

> ✅ (Open to feedback on this) I transitioned to using absolute dates for better formatting, but open to feedback on this!

My proposal is to make a mix of the relative dates in the blame column for easier scanning and relative comparisons between the rows and to show the full timestamp in the overlay since we have more space there and need precision (the relative date was already shown before).

**TL;DR: Awaiting feedback**

## Test plan

Screenshot 2022-09-06 at 12 25 40

## App preview:

- [Web](https://sg-web-ps-tweak-git-blame-uiux.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-kxqdhnxkld.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by jjinnii** [comment]
> I suppose this is already how it works (because of the video) but I have a hard time checking this

I am unable to get the Codecov extension running either, I'm going off of the video 😓 But in [Tara's Loom](https://www.loom.com/share/44bb0d3ec655455a9a730ffd4a61fd71), you can see that it is not the default behavior for blame to take precedence. 11 seconds into the video, you can see that the blame row highlight is green, because the Codecov decorations are taking precedence! 

> However, I think this is less of an issue after one of the other changes below that changes the position of the git blame popup. Since the popup is now no longer shifted to the right, you don't need to move the mouse outside the blame column but simply upwards/downwards which prevents any interferences with the code intel popup.

I agree that better anchoring the popover reduces the likelihood that users will experience this poor UX when trying to click into the blame popover. However, I still think there's a [weird latency](https://www.loom.com/share/da9d5d41d3f74ab2bdaea52be2969f86) when a user navigates their cursor outside of the blame column, irregardless of the code intel popover. And like you mentioned, this is not the case when going from the file blob into the blame column. 

TL;DR I still think it's an issue, but less so since we've fixed the positioning of the blame popover. If the effort to fix this latency is non-trivial, we can create a follow-up issue for it for post 4.0.

> I fixed this as good as possible. I needed to make changes to the paddings and the content text size. See the screenshot below.

It's hard to tell from just a screenshot of the blame popover, I might just have to double check this once it's been merged to the feature flag! 

> Are you specifically talking about an animation here or what do you refer to? The lines already show an underline when hovered so I’m not quite sure what is missing!

Yup you're right, it's already there! Sorry about that, I don't know why I didn't see it yesterday for some reason 😄 

> My proposal is to make a mix of the relative dates in the blame column for easier scanning and relative comparisons between the rows and to show the full timestamp in the overlay since we have more space there and need precision (the relative date was already shown before).

👍  I like this approach!

**Comment by jjinnii** [comment]
> Is there anything I can do to help you verify this before we merge or are you fine with what you saw and we just ship and and see?

Fine with shipping and seeing! The improvements to the popover positioning + your improvements are definitely sufficient for first release, any additional improvements would be nice to haves anyway 😄 

> Yes I see what you mean now. I think I'll leave this as a small fix for @taras-yemets tomorrow who can probably reproduce this easily (as he created the video already) okay?

Yes please tag me in the issue somewhere so I can follow along!

**Comment by taras-yemets** [comment]
> I am unable to get the Codecov extension running either, I'm going off of the video 😓 But in [Tara's Loom](https://www.loom.com/share/44bb0d3ec655455a9a730ffd4a61fd71), you can see that it is not the default behavior for blame to take precedence. 11 seconds into the video, you can see that the blame row highlight is green, because the Codecov decorations are taking precedence!

[Replied](https://github.com/sourcegraph/sourcegraph/issues/41314#issuecomment-1238974257) in the original issue.

**Review by taras-yemets — APPROVED** [review]
Awesome! Thank you Philipp!
